### PR TITLE
Analyzer: Fix virtual columns in StorageMerge

### PR DESCRIPTION
**Pull request by novikd**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 1a47682c12a45acb888c89974302ec35bee5aaed with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61298/1a47682c12a45acb888c89974302ec35bee5aaed/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Storages/StorageMerge.cpp [hunk 1]
```diff
@@ -1493,14 +1546,16 @@ void ReadFromMerge::convertAndFilterSourceStream(
...
-        && (processed_stage != QueryProcessingStage::FetchColumns || dynamic_cast<const StorageDistributed *>(&snapshot->storage) != nullptr))
+        && (processed_stage == QueryProcessingStage::FetchColumns && dynamic_cast<const StorageDistributed *>(&snapshot->storage) != nullptr))
         convert_actions_match_columns_mode = ActionsDAG::MatchColumnsMode::Position;
 
     if (row_policy_data_opt)
     {
         row_policy_data_opt->addFilterTransform(builder);
     }
 
+    LOG_DEBUG(&Poco::Logger::get("convertAndFilterSourceStream"), "SOURCE:\n{}\nRESULT:\n{}", builder.getHeader().dumpStructure(), header.dumpStructure());
```
Comment (alexey-milovidov) [inline src/Storages/StorageMerge.cpp:1557]: Remove before merge.

File: src/Storages/StorageMerge.cpp [hunk 2]
```diff
@@ -1047,40 +1058,80 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
 
         Block pipe_header = builder->getHeader();
 
-        if (has_database_virtual_column && common_header.has("_database") && !pipe_header.has("_database"))
+        LOG_DEBUG(&Poco::Logger::get("createSources"), "Processed:{}\nStorage:{}", toString(processed_stage), toString(storage_stage));
```
Comment (alexey-milovidov) [inline src/Storages/StorageMerge.cpp:1061]: Remove before merge.

File: src/Storages/StorageMerge.cpp [hunk 3]
```diff
@@ -798,10 +803,13 @@ QueryTreeNodePtr replaceTableExpressionAndRemoveJoin(
     const ContextPtr & context,
     const Names & required_column_names)
 {
+    LOG_DEBUG(&Poco::Logger::get("replaceTableExpressionAndRemoveJoin"), "BEFORE:\n{}", query->dumpTree());
```
Comment (alexey-milovidov) [inline src/Storages/StorageMerge.cpp:806]: .

File: src/Storages/StorageMerge.cpp [hunk 4]
```diff
@@ -798,10 +803,13 @@ QueryTreeNodePtr replaceTableExpressionAndRemoveJoin(
...
     auto * query_node = query->as<QueryNode>();
     auto join_tree_type = query_node->getJoinTree()->getNodeType();
     auto modified_query = query_node->cloneAndReplace(original_table_expression, replacement_table_expression);
 
+    LOG_DEBUG(&Poco::Logger::get("replaceTableExpressionAndRemoveJoin"), "AFTER:\n{}", modified_query->dumpTree());
```
Comment (alexey-milovidov) [inline src/Storages/StorageMerge.cpp:811]: .

**Review by alexey-milovidov — APPROVED** [review]

**Comment by antonio2368** [comment]
@alexey-milovidov 
- you didn't let CI finish
- the test that managed to fail in time was related to these changes

:disappointed:

**Comment by clickgapai** [comment]
Hi @novikd @alexey-milovidov — while reviewing this PR I found the following:

- **Coverage gap:** `merge() AS alias` with alias-qualified `_table`/`_database` virtual columns has zero coverage — https://github.com/ClickHouse/ClickHouse/pull/103923

Could you review the test PR and add the `can be tested` label if it looks good?
Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**